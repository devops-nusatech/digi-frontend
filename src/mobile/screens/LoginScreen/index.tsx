import classnames from 'classnames';
import * as React from 'react';
import { useIntl } from 'react-intl';
import { useSelector } from 'react-redux';
import { useHistory } from 'react-router';
import { selectLoginRequire2FA } from '../../../modules/user/auth';
import { LoginScreen } from '../../../screens/LoginScreen';
import { Modal } from '../../components';

const LoginMobileScreen: React.FC = () => {
   const require2FA = useSelector(selectLoginRequire2FA);
   const history = useHistory();
   const intl = useIntl();
   const title = require2FA
      ? { id: 'page.mobile.login.kyc.header' }
      : { id: 'page.body.landing.header.button2' };
   const className = classnames({
      'cr-mobile-login': !require2FA,
      'cr-mobile-kyc': require2FA,
   });

   return (
      <div className={className}>
         <Modal
            isOpen={true}
            onClose={() => history.push('/trading')}
            onBack={() => !require2FA && history.push('/register')}
            backTitle={intl.formatMessage({
               id: 'page.body.landing.header.button3',
            })}
            title={intl.formatMessage(title)}>
            <LoginScreen />
         </Modal>
      </div>
   );
};

export { LoginMobileScreen };
