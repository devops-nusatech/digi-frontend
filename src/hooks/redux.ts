import { shallowEqual, useSelector } from 'react-redux';
import { RootState } from 'modules';

export const useReduxSelector = <T>(selector: (state: RootState) => T): T =>
   useSelector(selector, shallowEqual);
