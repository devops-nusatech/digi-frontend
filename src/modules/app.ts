import { combineReducers } from 'redux';
import { airdropsReducer } from './public/airdrops';
import { alertReducer } from './public/alert';
import { blockchainsReducer } from './public/blockchains';
import { blocklistAccessReducer } from './public/blocklistAccess';
import { configsReducer } from './public/configs';
import { currenciesReducer } from './public/currencies';
import { customizationReducer } from './public/customization';
import { errorHandlerReducer } from './public/errorHandler';
import { changeColorThemeReducer } from './public/globalSettings';
import { gridLayoutReducer } from './public/gridLayout';
import { changeLanguageReducer } from './public/i18n';
import { klineReducer } from './public/kline';
import { marketsReducer } from './public/markets';
import { memberLevelsReducer } from './public/memberLevels';
import { membershipsReducer } from './public/memberships';
import { newsReducer } from './public/news';
import { news2Reducer } from './public/news2';
import { depthReducer, incrementDepthReducer, orderBookReducer } from './public/orderBook';
import { rangerReducer } from './public/ranger/reducer';
import { recentTradesReducer } from './public/recentTrades';
import { tradingFeesReducer } from './public/tradingFees';
import { withdrawLimitsReducer } from './public/withdrawLimits';
import { apiKeysReducer } from './user/apiKeys';
import { authReducer } from './user/auth';
import { beneficiariesReducer } from './user/beneficiaries';
import { getGeetestCaptchaReducer } from './user/captcha';
import { customizationUpdateReducer } from './user/customization';
import { documentationReducer } from './user/documentation';
import { sendEmailVerificationReducer } from './user/emailVerification';
import { groupMemberReducer } from './user/groupMember';
import { historyReducer } from './user/history';
import {
   addressesReducer,
   documentsReducer,
   identityReducer,
   labelReducer,
   phoneReducer,
} from './user/kyc';
import { openOrdersReducer } from './user/openOrders';
import { ordersReducer } from './user/orders';
import { ordersHistoryReducer } from './user/ordersHistory';
import { passwordReducer } from './user/password';
import { profileReducer } from './user/profile';
import { tierReducer } from './user/tier';
import { transactionsReducer } from './user/transactions';
import { transferReducer } from './user/transfers';
import { userActivityReducer } from './user/userActivity';
import { walletsReducer } from './user/wallets';
import { withdrawLimitReducer } from './user/withdrawLimit';

export const publicReducer = combineReducers({
   airdrops: airdropsReducer,
   alerts: alertReducer,
   blockchains: blockchainsReducer,
   blocklistAccess: blocklistAccessReducer,
   colorTheme: changeColorThemeReducer,
   configs: configsReducer,
   currencies: currenciesReducer,
   customization: customizationReducer,
   errorHandler: errorHandlerReducer,
   rgl: gridLayoutReducer,
   i18n: changeLanguageReducer,
   kline: klineReducer,
   markets: marketsReducer,
   memberLevels: memberLevelsReducer,
   memberships: membershipsReducer,
   news: newsReducer,
   news2: news2Reducer,
   orderBook: orderBookReducer,
   depth: depthReducer,
   incrementDepth: incrementDepthReducer,
   ranger: rangerReducer,
   recentTrades: recentTradesReducer,
   tradingFees: tradingFeesReducer,
   withdrawLimits: withdrawLimitsReducer,
});

export const userReducer = combineReducers({
   addresses: addressesReducer,
   apiKeys: apiKeysReducer,
   auth: authReducer,
   beneficiaries: beneficiariesReducer,
   captcha: getGeetestCaptchaReducer,
   customizationUpdate: customizationUpdateReducer,
   documentation: documentationReducer,
   history: historyReducer,
   documents: documentsReducer,
   identity: identityReducer,
   label: labelReducer,
   phone: phoneReducer,
   openOrders: openOrdersReducer,
   orders: ordersReducer,
   ordersHistory: ordersHistoryReducer,
   password: passwordReducer,
   profile: profileReducer,
   tier: tierReducer,
   groupMember: groupMemberReducer,
   transactions: transactionsReducer,
   transfer: transferReducer,
   sendEmailVerification: sendEmailVerificationReducer,
   userActivity: userActivityReducer,
   wallets: walletsReducer,
   withdrawLimit: withdrawLimitReducer,
});
