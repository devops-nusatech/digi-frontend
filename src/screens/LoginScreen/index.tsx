import cx from 'classnames';
import * as React from 'react';
import { injectIntl } from 'react-intl';
import {
   connect,
   MapDispatchToPropsFunction,
   MapStateToProps,
} from 'react-redux';
import { RouterProps } from 'react-router';
import { withRouter } from 'react-router-dom';
import { compose } from 'redux';
import { IntlProps } from '../..';
import { captchaLogin, captchaType } from '../../api';
import { Captcha, LoginComponent, TwoFactorAuth } from '../../components';
import {
   EMAIL_REGEX,
   ERROR_EMPTY_PASSWORD,
   ERROR_INVALID_EMAIL,
   setDocumentTitle,
} from '../../helpers';
import {
   GeetestCaptchaResponse,
   GeetestCaptchaV4Response,
   resetCaptchaState,
   RootState,
   selectAlertState,
   selectCaptchaResponse,
   selectGeetestCaptchaSuccess,
   selectRecaptchaSuccess,
   selectLoginError,
   selectLoginRequire2FA,
   selectregisterRequireVerification,
   selectUserFetching,
   selectUserLoggedIn,
   login,
   loginError,
   loginRequire2FA,
   registerRequireVerification,
} from '../../modules';
import { CommonError } from '../../modules/types';

interface ReduxProps {
   error?: CommonError;
   isLoggedIn: boolean;
   loading?: boolean;
   require2FA?: boolean;
   requireEmailVerification?: boolean;
   captcha_response?:
      | string
      | GeetestCaptchaResponse
      | GeetestCaptchaV4Response;
   reCaptchaSuccess: boolean;
   geetestCaptchaSuccess: boolean;
}

interface DispatchProps {
   login: typeof login;
   loginError: typeof loginError;
   loginRequire2FA: typeof loginRequire2FA;
   resetCaptchaState: typeof resetCaptchaState;
   registerRequireVerification: typeof registerRequireVerification;
}

interface LoginState {
   email: string;
   emailError: string;
   emailFocused: boolean;
   password: string;
   passwordError: string;
   passwordFocused: boolean;
   otpCode: string;
   error2fa: string;
   codeFocused: boolean;
}

type Props = ReduxProps & DispatchProps & RouterProps & IntlProps;

class Login extends React.Component<Props, LoginState> {
   public state = {
      email: '',
      emailError: '',
      emailFocused: false,
      password: '',
      passwordError: '',
      passwordFocused: false,
      otpCode: '',
      error2fa: '',
      codeFocused: false,
   };

   public componentDidMount() {
      setDocumentTitle('Sign In');
      this.props.loginError({ code: 0, message: [''] });
      this.props.registerRequireVerification({ requireVerification: false });
   }

   public componentWillReceiveProps(nextProps: Props) {
      const { email } = this.state;

      if (!this.props.isLoggedIn && nextProps.isLoggedIn) {
         this.props.resetCaptchaState();
         this.props.history.push('/wallets', { email });
      }

      if (nextProps.requireEmailVerification) {
         this.props.history.push('/email-verification', { email });
      }
   }

   public componentWillUnmount() {
      this.props.resetCaptchaState();
   }

   public renderCaptcha = () => {
      const { error } = this.props;

      return <Captcha error={error} />;
   };

   public render() {
      const { loading, require2FA } = this.props;

      const className = cx('pg-login-screen__container', { loading });

      return (
         <div className="pg-login-screen">
            <div className={className}>
               {require2FA ? this.render2FA() : this.renderLoginForm()}
            </div>
         </div>
      );
   }

   private renderLoginForm = () => {
      const {
         loading,
         captcha_response,
         reCaptchaSuccess,
         geetestCaptchaSuccess,
      } = this.props;
      const {
         email,
         emailError,
         emailFocused,
         password,
         passwordError,
         passwordFocused,
      } = this.state;

      return (
         <LoginComponent
            email={email}
            emailError={emailError}
            emailFocused={emailFocused}
            emailPlaceholder={this.props.intl.formatMessage({
               id: 'page.header.login.email',
            })}
            password={password}
            passwordError={passwordError}
            passwordFocused={passwordFocused}
            passwordPlaceholder={this.props.intl.formatMessage({
               id: 'page.header.login.password',
            })}
            labelLogin={this.props.intl.formatMessage({
               id: 'page.header.login',
            })}
            labelregister={this.props.intl.formatMessage({
               id: 'page.header.register',
            })}
            emailLabel={this.props.intl.formatMessage({
               id: 'page.header.login.email',
            })}
            passwordLabel={this.props.intl.formatMessage({
               id: 'page.header.login.password',
            })}
            receiveConfirmationLabel={this.props.intl.formatMessage({
               id: 'page.header.login.receiveConfirmation',
            })}
            forgotPasswordLabel={this.props.intl.formatMessage({
               id: 'page.header.login.forgotPassword',
            })}
            isLoading={loading}
            onForgotPassword={this.forgotPassword}
            onregister={this.handleregister}
            onLogin={this.handleLogin}
            handleChangeFocusField={this.handleFieldFocus}
            isFormValid={this.validateForm}
            refreshError={this.refreshError}
            changeEmail={this.handleChangeEmailValue}
            changePassword={this.handleChangePasswordValue}
            captchaType={captchaType()}
            renderCaptcha={this.renderCaptcha()}
            reCaptchaSuccess={reCaptchaSuccess}
            geetestCaptchaSuccess={geetestCaptchaSuccess}
            captcha_response={captcha_response}
         />
      );
   };

   private render2FA = () => {
      const { loading } = this.props;
      const { otpCode, error2fa, codeFocused } = this.state;

      return (
         <TwoFactorAuth
            isLoading={loading}
            onSubmit={this.handle2FALogin}
            title={this.props.intl.formatMessage({ id: 'page.password2fa' })}
            label={this.props.intl.formatMessage({
               id: 'page.body.wallets.tabs.withdraw.content.code2fa',
            })}
            buttonLabel={this.props.intl.formatMessage({
               id: 'page.header.login',
            })}
            message={this.props.intl.formatMessage({
               id: 'page.password2fa.message',
            })}
            codeFocused={codeFocused}
            otpCode={otpCode}
            error={error2fa}
            handleOtpCodeChange={this.handleChangeOtpCode}
            handleChangeFocusField={this.handle2faFocus}
            handleClose2fa={this.handleClose}
         />
      );
   };

   private refreshError = () => {
      this.setState({
         emailError: '',
         passwordError: '',
      });
   };

   private handleChangeOtpCode = (value: string) => {
      this.setState({
         error2fa: '',
         otpCode: value,
      });
   };

   private handleLogin = () => {
      const { email, password } = this.state;
      const { captcha_response } = this.props;

      if (captchaType() !== 'none' && captchaLogin()) {
         this.props.login({ email, password, captcha_response });
      } else {
         this.props.login({ email, password });
      }
   };

   private handle2FALogin = () => {
      const { email, password, otpCode } = this.state;
      const { captcha_response } = this.props;

      if (!otpCode) {
         this.setState({
            error2fa: 'Please enter 2fa code',
         });
      } else if (captchaType() !== 'none' && captchaLogin()) {
         this.props.login({
            email,
            password,
            captcha_response,
            otp_code: otpCode,
         });
      } else {
         this.props.login({ email, password, otp_code: otpCode });
      }
   };

   private handleregister = () => {
      this.props.history.push('/register');
   };

   private forgotPassword = () => {
      this.props.history.push('/forgot_password');
   };

   private handleFieldFocus = (field: string) => {
      switch (field) {
         case 'email':
            this.setState(prev => ({
               emailFocused: !prev.emailFocused,
            }));
            break;
         case 'password':
            this.setState(prev => ({
               passwordFocused: !prev.passwordFocused,
            }));
            break;
         default:
            break;
      }
   };

   private handle2faFocus = () => {
      this.setState(prev => ({
         codeFocused: !prev.codeFocused,
      }));
   };

   private validateForm = () => {
      const { email, password } = this.state;
      const isEmailValid = email.match(EMAIL_REGEX);

      if (!isEmailValid) {
         this.setState({
            emailError: this.props.intl.formatMessage({
               id: ERROR_INVALID_EMAIL,
            }),
            passwordError: '',
         });

         return;
      }
      if (!password) {
         this.setState({
            emailError: '',
            passwordError: this.props.intl.formatMessage({
               id: ERROR_EMPTY_PASSWORD,
            }),
         });

         return;
      }
   };

   private handleChangeEmailValue = (value: string) => {
      this.setState({
         email: value,
      });
   };

   private handleChangePasswordValue = (value: string) => {
      this.setState({
         password: value,
      });
   };

   private handleClose = () => {
      this.props.loginRequire2FA({ require2fa: false });
   };
}

const mapStateToProps: MapStateToProps<ReduxProps, {}, RootState> = state => ({
   alert: selectAlertState(state),
   error: selectLoginError(state),
   isLoggedIn: selectUserLoggedIn(state),
   loading: selectUserFetching(state),
   require2FA: selectLoginRequire2FA(state),
   requireEmailVerification: selectregisterRequireVerification(state),
   captcha_response: selectCaptchaResponse(state),
   reCaptchaSuccess: selectRecaptchaSuccess(state),
   geetestCaptchaSuccess: selectGeetestCaptchaSuccess(state),
});

const mapDispatchToProps: MapDispatchToPropsFunction<
   DispatchProps,
   {}
> = dispatch => ({
   login: data => dispatch(login(data)),
   loginError: error => dispatch(loginError(error)),
   loginRequire2FA: payload => dispatch(loginRequire2FA(payload)),
   resetCaptchaState: () => dispatch(resetCaptchaState()),
   registerRequireVerification: data =>
      dispatch(registerRequireVerification(data)),
});

export const LoginScreen = compose(
   injectIntl,
   withRouter,
   connect(mapStateToProps, mapDispatchToProps)
)(Login) as React.ComponentClass;
