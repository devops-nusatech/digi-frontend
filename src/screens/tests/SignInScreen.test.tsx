import { shallow } from 'enzyme';
import * as React from 'react';
import { connect, Provider } from 'react-redux';
import { createStore } from 'redux';
import { IntlProps } from '../../';
import { rootReducer } from '../../modules';
import { LoginScreen } from '../LoginScreen';

const store = createStore(rootReducer);
const Identity = connect()(LoginScreen);

const setup = (props: Partial<IntlProps> = {}) =>
   shallow(
      <Provider store={store}>
         <Identity />
      </Provider>
   );

describe('LoginScreen', () => {
   const wrapper = setup();

   it('should render', () => {
      expect(wrapper).toMatchSnapshot();
   });
});
