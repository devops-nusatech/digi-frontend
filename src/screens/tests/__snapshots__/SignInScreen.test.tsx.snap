// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen should render 1`] = `ShallowWrapper {}`;
